File: ecosystems/rubygems/lib/aitextcleaner/formatting_residue.rb
require "cgi"

module Aitextcleaner
  module FormattingResidue
    BLOCK_TAG = /<\/?(?:address|article|blockquote|br|div|li|p|pre|section)\b[^>]*>/i
    HTML_TAG = /<[^>]*>/

    module_function

    def clean(text)
      raise TypeError, "text must be a string" unless text.is_a?(String)

      cleaned = text.gsub(BLOCK_TAG, "\n").gsub(HTML_TAG, "")
      cleaned = CGI.unescapeHTML(cleaned)
      cleaned = cleaned.gsub("“", '"').gsub("”", '"').gsub("‘", "'").gsub("’", "'")
        .gsub("–", "-").gsub("—", "-").gsub("…", "...")
      cleaned.gsub(/[ \t]+\n/, "\n").gsub(/\n{3,}/, "\n\n").strip
    end
  end
end

File: ecosystems/nuget/src/FormattingResidue.cs
using System;
using System.Net;
using System.Text.RegularExpressions;

namespace AiTextCleaner.FormattingResidue;

public static class FormattingResidue
{
    private static readonly Regex BlockTag = new(
        @"</?(?:address|article|blockquote|br|div|li|p|pre|section)\b[^>]*>",
        RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex HtmlTag = new("<[^>]*>", RegexOptions.Compiled);

    public static string Clean(string text)
    {
        ArgumentNullException.ThrowIfNull(text);
        var cleaned = BlockTag.Replace(text, "\n");
        cleaned = HtmlTag.Replace(cleaned, string.Empty);
        cleaned = WebUtility.HtmlDecode(cleaned)
            .Replace('“', '"')
            .Replace('”', '"')
            .Replace('‘', '\'')
            .Replace('’', '\'')
            .Replace('–', '-')
            .Replace('—', '-')
            .Replace("…", "...");
        cleaned = Regex.Replace(cleaned, "[ \\t]+\\n", "\\n");
        cleaned = Regex.Replace(cleaned, "\\n{3,}", "\\n\\n");
        return cleaned.Trim();
    }
}
